test/data/CONFIG.TXT
; Firmware version v20210520

; For information on configuring FlySight, please go to
;     http://flysight.ca/wiki

; GPS settings

Model:     7     ; Dynamic model
                 ;   0 = Portable
                 ;   2 = Stationary
                 ;   3 = Pedestrian
                 ;   4 = Automotive
                 ;   5 = Sea
                 ;   6 = Airborne with < 1 G acceleration
                 ;   7 = Airborne with < 2 G acceleration
                 ;   8 = Airborne with < 4 G acceleration
Rate:      100   ; Measurement rate (ms)

; Tone settings

Limits:    1     ; Behaviour when outside bounds
                 ;   0 = No tone
                 ;   1 = Min/max tone
                 ;   2 = Chirp up/down
                 ;   3 = Chirp down/up
Volume:    6     ; 0 (min) to 8 (max)

; Rate settings

Min_Rate:  100   ; Minimum rate (Hz * 100)
Max_Rate:  500   ; Maximum rate (Hz * 100)
Flatline:  0     ; Flatline at minimum rate
                 ;   0 = No
                 ;   1 = Yes

; Thresholds

V_Thresh:  1000  ; Minimum vertical speed for tone (cm/s)
H_Thresh:  0     ; Minimum horizontal speed for tone (cm/s)

; Acro settings

Acro_Win:           -1   ; Acro window, meters below exit

; Exit detection

Exit_Min_Alt:       -1   ; Do not detect an exit if below this altitude in meters
                         ; (negative to disable exit detection)
Exit_Exit_Alt:      -1   ; Do not detect an exit if above this altitude in meters
                         ; (negative to disable this check)
Exit_Num_Down:      5    ; Consecutive points down to indicate an exit
Exit_Num_Up:        50   ; Consecutive points up reset the exit altitude
Exit_Down_Thresh:   800  ; Speed (cm/s) to indicate down (positive) (initialize exit altitude)
Exit_Up_Thresh:     -800 ; Speed (cm/s) to indicate up (negative) (reset exit altitude)

; Performance lane

Lane_Width:          600 ; Width of lane in meters
Lane_Num_LEDs:        14 ; Number of leds for lane indicator
Lane_LED_Offset:       3 ; First LED in display
Time_After_Exit:      -1 ; Lane start time after ext (ms) (negative to disable)
Reference_Lat:         0 ; Reference latitude (deg * 1e7)
Reference_Lon:         0 ; Reference longitude (deg * 1e7)

; Enable Performance lane

Exit_Min_Alt:        2286 ; Do not detect an exit if below this altitude in meters
Exit_Exit_Alt:       3353 ; Do not detect an exit if above this altitude in meters
Time_After_Exit:    10000 ; Lane start time after ext (ms) (negative to disable)

; ref_name: s2n_1
;Reference_Lat:  326680550
;Reference_Lon: -1116970660

; ref_name: s2n_2
;Reference_Lat:  326648030
;Reference_Lon: -1116883540

; ref_name: s2n_3
;Reference_Lat:  326707820
;Reference_Lon: -1116804790

; ref_name: s2n_4
;Reference_Lat:  326691750
;Reference_Lon: -1116722180

; ref_name: spa_fbo
;Reference_Lat:  411449087
;Reference_Lon: -801676433

; ref_name: spa_landmark
;Reference_Lat:  411359021
;Reference_Lon: -801791902

; ref_name: runway
;Reference_Lat:  411458368
;Reference_Lon: -801604920

; ref_name: runway
Reference_LatF:  41.1458368
Reference_LonF: -80.1604920

; Miscellaneous

Use_SAS:   0     ; Use skydiver's airspeed
                 ;   0 = No
                 ;   1 = Yes
TZ_Offset: 0     ; Timezone offset of output files in seconds
                 ;   -14400 = UTC-4 (EDT)
                 ;   -18000 = UTC-5 (EST, CDT)
                 ;   -21600 = UTC-6 (CST, MDT)
                 ;   -25200 = UTC-7 (MST, PDT)
                 ;   -28800 = UTC-8 (PST)

; Initialization

Init_Mode: 0     ; When the FlySight is powered on
                 ;   0 = Do nothing
                 ;   1 = Test speech mode
                 ;   2 = Play file
Init_File: 0     ; File to be played

; Alarm settings

; WARNING: GPS measurements depend on very weak signals
;          received from orbiting satellites. As such, they
;          are prone to interference, and should NEVER be
;          relied upon for life saving purposes.

;          UNDER NO CIRCUMSTANCES SHOULD THESE ALARMS BE
;          USED TO INDICATE DEPLOYMENT OR BREAKOFF ALTITUDE.

; NOTE:    Alarm elevations are given in meters above ground
;          elevation, which is specified in DZ_Elev.

Win_Above:    20 ; Window above each alarm (m)
Win_Below:    20 ; Window below each alarm (m)
DZ_Elev:     418 ; Ground elevation (m above sea level) - Skydive Pennsylvania
;DZ_Elev:     338 ; Ground elevation (m above sea level) - Cleveland Skydiving Center
;DZ_Elev:     472 ; Ground elevation (m above sea level) - Sawtooth
;DZ_Elev:     24 ; Ground elevation (m above sea level) - Skydive Deland
;DZ_Elev:     188 ; Ground elevation (m above sea level) - Skydive Chicago

; Altitude mode settings

; WARNING: GPS measurements depend on very weak signals
;          received from orbiting satellites. As such, they
;          are prone to interference, and should NEVER be
;          relied upon for life saving purposes.

;          UNDER NO CIRCUMSTANCES SHOULD ALTITUDE MODE BE
;          USED TO INDICATE DEPLOYMENT OR BREAKOFF ALTITUDE.

; NOTE:    Altitude is given relative to ground elevation,
;          which is specified in DZ_Elev. Altitude mode will
;          not function below 1500 m above ground.

Alt_Units:     1 ; Altitude units
                 ;   0 = m
                 ;   1 = ft
Alt_Step:      0 ; Altitude between announcements
                 ;   0 = No altitude

Wind_hMSL:    1336687 ; 918.687 (m agl)
Wind_velE:       9177 ; 33.038 (kmph)
Wind_velN:       3673 ; 13.224 (kmph)

Wind_hMSL:    1643127 ; 1225.128 (m agl)
Wind_velE:       9909 ; 35.676 (kmph)
Wind_velN:       2602 ; 9.367 (kmph)

Wind_hMSL:    1991449 ; 1573.449 (m agl)
Wind_velE:      10592 ; 38.131 (kmph)
Wind_velN:       1336 ; 4.810 (kmph)

Wind_hMSL:    2386075 ; 1968.076 (m agl)
Wind_velE:      12213 ; 43.970 (kmph)
Wind_velN:        790 ; 2.847 (kmph)

Wind_hMSL:    2832282 ; 2414.283 (m agl)
Wind_velE:      14124 ; 50.847 (kmph)
Wind_velN:       1145 ; 4.126 (kmph)

Wind_hMSL:    3336756 ; 2918.756 (m agl)
Wind_velE:      15877 ; 57.159 (kmph)
Wind_velN:       1725 ; 6.213 (kmph)

Wind_hMSL:    3906311 ; 3488.312 (m agl)
Wind_velE:      16723 ; 60.205 (kmph)
Wind_velN:       1625 ; 5.850 (kmph)

Wind_hMSL:    4533619 ; 4115.620 (m agl)
Wind_velE:      16788 ; 60.439 (kmph)
Wind_velN:        989 ; 3.564 (kmph)


src/Config.cpp
/***************************************************************************
**                                                                        **
**  FlySight firmware                                                     **
**  Copyright 2018 Michael Cooper, Will Glynn                             **
**                                                                        **
**  This program is free software: you can redistribute it and/or modify  **
**  it under the terms of the GNU General Public License as published by  **
**  the Free Software Foundation, either version 3 of the License, or     **
**  (at your option) any later version.                                   **
**                                                                        **
**  This program is distributed in the hope that it will be useful,       **
**  but WITHOUT ANY WARRANTY; without even the implied warranty of        **
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         **
**  GNU General Public License for more details.                          **
**                                                                        **
**  You should have received a copy of the GNU General Public License     **
**  along with this program.  If not, see <http://www.gnu.org/licenses/>. **
**                                                                        **
****************************************************************************
**  Contact: Michael Cooper                                               **
**  Website: http://flysight.ca/                                          **
****************************************************************************/

#include <avr/eeprom.h>
#include <avr/pgmspace.h>

#include <stdlib.h>
#include <string.h>

#include "Board/LEDs.h"
#include "FatFS/ff.h"
#include "Config.h"
#include "Debug.h"
#include "Log.h"
#include "Main.h"
#include "Tone.h"
#include "UBX.h"
#include "Version.h"

#define FALSE 0
#define TRUE  (!FALSE)

#define CONFIG_FIRST_ALARM  (1 << 0)
#define CONFIG_FIRST_WINDOW (1 << 1)
#define CONFIG_FIRST_SPEECH (1 << 2)
#define CONFIG_FIRST_WIND   (1 << 3)

static const char Config_default[] PROGMEM = "\
; Firmware version " FLYSIGHT_VERSION "\r\n\
\r\n\
; For information on configuring FlySight, please go to\r\n\
;     http://flysight.ca/wiki\r\n\
\r\n\
; GPS settings\r\n\
\r\n\
Model:     7     ; Dynamic model\r\n\
                 ;   0 = Portable\r\n\
                 ;   2 = Stationary\r\n\
                 ;   3 = Pedestrian\r\n\
                 ;   4 = Automotive\r\n\
                 ;   5 = Sea\r\n\
                 ;   6 = Airborne with < 1 G acceleration\r\n\
                 ;   7 = Airborne with < 2 G acceleration\r\n\
                 ;   8 = Airborne with < 4 G acceleration\r\n\
Rate:      200   ; Measurement rate (ms)\r\n\
\r\n\
; Tone settings\r\n\
\r\n\
Mode:      2     ; Measurement mode\r\n\
                 ;   0 = Horizontal speed\r\n\
                 ;   1 = Vertical speed\r\n\
                 ;   2 = Glide ratio\r\n\
                 ;   3 = Inverse glide ratio\r\n\
                 ;   4 = Total speed\r\n\
                 ;   11 = Dive angle\r\n\
Min:       0     ; Lowest pitch value\r\n\
                 ;   cm/s        in Mode 0, 1, or 4\r\n\
                 ;   ratio * 100 in Mode 2 or 3\r\n\
                 ;   degrees     in Mode 11\r\n\
Max:       300   ; Highest pitch value\r\n\
                 ;   cm/s        in Mode 0, 1, or 4\r\n\
                 ;   ratio * 100 in Mode 2 or 3\r\n\
                 ;   degrees     in Mode 11\r\n\
Limits:    1     ; Behaviour when outside bounds\r\n\
                 ;   0 = No tone\r\n\
                 ;   1 = Min/max tone\r\n\
                 ;   2 = Chirp up/down\r\n\
                 ;   3 = Chirp down/up\r\n\
Volume:    6     ; 0 (min) to 8 (max)\r\n\
\r\n\
; Rate settings\r\n\
\r\n\
Mode_2:    9     ; Determines tone rate\r\n\
                 ;   0 = Horizontal speed\r\n\
                 ;   1 = Vertical speed\r\n\
                 ;   2 = Glide ratio\r\n\
                 ;   3 = Inverse glide ratio\r\n\
                 ;   4 = Total speed\r\n\
                 ;   8 = Magnitude of Value 1\r\n\
                 ;   9 = Change in Value 1\r\n\
                 ;   11 = Dive angle\r\n\
Min_Val_2: 300   ; Lowest rate value\r\n\
                 ;   cm/s          when Mode 2 = 0, 1, or 4\r\n\
                 ;   ratio * 100   when Mode 2 = 2 or 3\r\n\
                 ;   percent * 100 when Mode 2 = 9\r\n\
                 ;   degrees       when Mode 2 = 11\r\n\
Max_Val_2: 1500  ; Highest rate value\r\n\
                 ;   cm/s          when Mode 2 = 0, 1, or 4\r\n\
                 ;   ratio * 100   when Mode 2 = 2 or 3\r\n\
                 ;   percent * 100 when Mode 2 = 9\r\n\
                 ;   degrees       when Mode 2 = 11\r\n\
Min_Rate:  100   ; Minimum rate (Hz * 100)\r\n\
Max_Rate:  500   ; Maximum rate (Hz * 100)\r\n\
Flatline:  0     ; Flatline at minimum rate\r\n\
                 ;   0 = No\r\n\
                 ;   1 = Yes\r\n\
\r\n\
; Speech settings\r\n\
\r\n\
Sp_Rate:   0     ; Speech rate (s)\r\n\
                 ;   0 = No speech\r\n\
Sp_Volume: 8     ; 0 (min) to 8 (max)\r\n\
\r\n\
Sp_Mode:   2     ; Speech mode\r\n\
                 ;   0 = Horizontal speed\r\n\
                 ;   1 = Vertical speed\r\n\
                 ;   2 = Glide ratio\r\n\
                 ;   3 = Inverse glide ratio\r\n\
                 ;   4 = Total speed\r\n\
                 ;   11 = Dive angle\r\n\
                 ;   12 = Altitude above DZ_Elev\r\n\
Sp_Units:  1     ; Speech units\r\n\
                 ;   0 = km/h or m\r\n\
                 ;   1 = mph or feet\r\n\
Sp_Dec:    1     ; Speech precision\r\n\
                 ;   Altitude step in Mode 5\r\n\
                 ;   Decimal places in all other Modes\r\n\
\r\n\
; Thresholds\r\n\
\r\n\
V_Thresh:  1000  ; Minimum vertical speed for tone (cm/s)\r\n\
H_Thresh:  0     ; Minimum horizontal speed for tone (cm/s)\r\n\
\r\n\
; Miscellaneous\r\n\
\r\n\
Use_SAS:   1     ; Use skydiver's airspeed\r\n\
                 ;   0 = No\r\n\
                 ;   1 = Yes\r\n\
TZ_Offset: 0     ; Timezone offset of output files in seconds\r\n\
                 ;   -14400 = UTC-4 (EDT)\r\n\
                 ;   -18000 = UTC-5 (EST, CDT)\r\n\
                 ;   -21600 = UTC-6 (CST, MDT)\r\n\
                 ;   -25200 = UTC-7 (MST, PDT)\r\n\
                 ;   -28800 = UTC-8 (PST)\r\n\
\r\n\
; Initialization\r\n\
\r\n\
Init_Mode: 0     ; When the FlySight is powered on\r\n\
                 ;   0 = Do nothing\r\n\
                 ;   1 = Test speech mode\r\n\
                 ;   2 = Play file\r\n\
Init_File: 0     ; File to be played\r\n\
\r\n\
; Acro settings\r\n\
\r\n\
Acro_Win:           -1   ; Acro window, meters below exit\r\n\
\r\n\
; Exit detection\r\n\
\r\n\
Exit_Min_Alt:       -1   ; Do not detect an exit if below this altitude\r\n\
                         ; (negative to disable exit detection)\r\n\
Exit_Exit_Alt:      -1   ; Do not detect an exit if above this altitude\r\n\
                         ; (negative to disable this check)\r\n\
Exit_Num_Down:      5    ; Consecutive points down to indicate an exit\r\n\
Exit_Num_Up:        50   ; Consecutive points up reset the exit altitude\r\n\
Exit_Down_Thresh:   800  ; Speed (cm/s) to indicate down (positive) (initialize exit altitude)\r\n\
Exit_Up_Thresh:     -800 ; Speed (cm/s) to indicate up (negative) (reset exit altitude)\r\n\
\r\n\
; Performance lane\r\n\
\r\n\
Lane_Width:          600 ; Width of lane in meters\r\n\
Lane_Num_LEDs:        15 ; Number of leds for lane indicator\r\n\
Lane_LED_Offset:       0 ; First LED in display\r\n\
Time_After_Exit:      -1 ; Lane start time after ext (ms) (negative to disable)\r\n\
Num_Blink_LEDs:       -1 ; Number of leds to blink before exit detection\r\n\
Reference_Lat:         0 ; Reference latitude (deg * 1e7)\r\n\
Reference_Lon:         0 ; Reference longitude (deg * 1e7)\r\n\
Reference_LatF:      0.0 ; Reference latitude (deg)\r\n\
Reference_LonF:      0.0 ; Reference longitude (deg)\r\n\
\r\n\
; Alarm settings\r\n\
\r\n\
; WARNING: GPS measurements depend on very weak signals\r\n\
;          received from orbiting satellites. As such, they\r\n\
;          are prone to interference, and should NEVER be\r\n\
;          relied upon for life saving purposes.\r\n\
\r\n\
;          UNDER NO CIRCUMSTANCES SHOULD THESE ALARMS BE\r\n\
;          USED TO INDICATE DEPLOYMENT OR BREAKOFF ALTITUDE.\r\n\
\r\n\
; NOTE:    Alarm elevations are given in meters above ground\r\n\
;          elevation, which is specified in DZ_Elev.\r\n\
\r\n\
Win_Above:     0 ; Window above each alarm (m)\r\n\
Win_Below:     0 ; Window below each alarm (m)\r\n\
DZ_Elev:       0 ; Ground elevation (m above sea level)\r\n\
\r\n\
Alarm_Elev:    0 ; Alarm elevation (m above ground level)\r\n\
Alarm_Type:    0 ; Alarm type\r\n\
                 ;   0 = No alarm\r\n\
                 ;   1 = Beep\r\n\
                 ;   2 = Chirp up\r\n\
                 ;   3 = Chirp down\r\n\
                 ;   4 = Play file\r\n\
Alarm_File:    0 ; File to be played\r\n\
Alarm_Acro:    0 ; Alarm is height above bottom of acro window\r\n\
\r\n\
; Altitude mode settings\r\n\
\r\n\
; WARNING: GPS measurements depend on very weak signals\r\n\
;          received from orbiting satellites. As such, they\r\n\
;          are prone to interference, and should NEVER be\r\n\
;          relied upon for life saving purposes.\r\n\
\r\n\
;          UNDER NO CIRCUMSTANCES SHOULD ALTITUDE MODE BE\r\n\
;          USED TO INDICATE DEPLOYMENT OR BREAKOFF ALTITUDE.\r\n\
\r\n\
; NOTE:    Altitude is given relative to ground elevation,\r\n\
;          which is specified in DZ_Elev. Altitude mode will\r\n\
;          not function below 1500 m above ground.\r\n\
\r\n\
Alt_Units:     1 ; Altitude units\r\n\
                 ;   0 = m\r\n\
                 ;   1 = ft\r\n\
Alt_Step:      0 ; Altitude between announcements\r\n\
                 ;   0 = No altitude\r\n\
\r\n\
; Silence windows\r\n\
\r\n\
; NOTE:    Silence windows are given in meters above ground\r\n\
;          elevation, which is specified in DZ_Elev. Tones\r\n\
;          will be silenced during these windows and only\r\n\
;          alarms will be audible.\r\n\
\r\n\
Win_Top:       0 ; Silence window top (m)\r\n\
Win_Bottom:    0 ; Silence window bottom (m)\r\n";

static const char Config_Model[] PROGMEM      = "Model";
static const char Config_Rate[] PROGMEM       = "Rate";
static const char Config_Mode[] PROGMEM       = "Mode";
static const char Config_Min[] PROGMEM        = "Min";
static const char Config_Max[] PROGMEM        = "Max";
static const char Config_Limits[] PROGMEM     = "Limits";
static const char Config_Volume[] PROGMEM     = "Volume";
static const char Config_Mode_2[] PROGMEM     = "Mode_2";
static const char Config_Min_Val_2[] PROGMEM  = "Min_Val_2";
static const char Config_Max_Val_2[] PROGMEM  = "Max_Val_2";
static const char Config_Min_Rate[] PROGMEM   = "Min_Rate";
static const char Config_Max_Rate[] PROGMEM   = "Max_Rate";
static const char Config_Flatline[] PROGMEM   = "Flatline";
static const char Config_Sp_Mode[] PROGMEM    = "Sp_Mode";
static const char Config_Sp_Units[] PROGMEM   = "Sp_Units";
static const char Config_Sp_Rate[] PROGMEM    = "Sp_Rate";
static const char Config_Sp_Dec[] PROGMEM     = "Sp_Dec";
static const char Config_Sp_Volume[] PROGMEM  = "Sp_Volume";
static const char Config_V_Thresh[] PROGMEM   = "V_Thresh";
static const char Config_H_Thresh[] PROGMEM   = "H_Thresh";
static const char Config_Use_SAS[] PROGMEM    = "Use_SAS";
static const char Config_Window[] PROGMEM     = "Window";
static const char Config_Window_Above[] PROGMEM = "Win_Above";
static const char Config_Window_Below[] PROGMEM = "Win_Below";
static const char Config_DZ_Elev[] PROGMEM    = "DZ_Elev";
static const char Config_Alarm_Elev[] PROGMEM = "Alarm_Elev";
static const char Config_Alarm_Type[] PROGMEM = "Alarm_Type";
static const char Config_Alarm_File[] PROGMEM = "Alarm_File";
static const char Config_Alarm_Acro[] PROGMEM = "Alarm_Acro";
static const char Config_Wind_hMSL[] PROGMEM  = "Wind_hMSL";
static const char Config_Wind_velN[] PROGMEM  = "Wind_velN";
static const char Config_Wind_velE[] PROGMEM  = "Wind_velE";
static const char Config_TZ_Offset[] PROGMEM  = "TZ_Offset";
static const char Config_Init_Mode[] PROGMEM  = "Init_Mode";
       const char Config_Init_File[] PROGMEM  = "Init_File";
static const char Config_Win_Top[] PROGMEM    = "Win_Top";
static const char Config_Win_Bottom[] PROGMEM = "Win_Bottom";
static const char Config_Alt_Units[] PROGMEM  = "Alt_Units";
static const char Config_Alt_Step[] PROGMEM   = "Alt_Step";
static const char Config_Acro_Win[] PROGMEM   = "Acro_Win";
static const char Config_Exit_Min_Alt[] PROGMEM = "Exit_Min_Alt";
static const char Config_Exit_Exit_Alt[] PROGMEM = "Exit_Exit_Alt";
static const char Config_Exit_Num_Down[] PROGMEM = "Exit_Num_Down";
static const char Config_Exit_Num_Up[] PROGMEM = "Exit_Num_Up";
static const char Config_Exit_Down_Thresh[] PROGMEM = "Exit_Down_Thresh";
static const char Config_Exit_Up_Thresh[] PROGMEM = "Exit_Up_Thresh";
static const char Config_Lane_Width[] PROGMEM = "Lane_Width";
static const char Config_Lane_Num_LEDs[] PROGMEM = "Lane_Num_LEDs";
static const char Config_Lane_LED_Offset[] PROGMEM = "Lane_LED_Offset";
static const char Config_Num_Blink_LEDs[] PROGMEM = "Num_Blink_LEDs";
static const char Config_Time_After_Exit[] PROGMEM = "Time_After_Exit";
static const char Config_Reference_Lat[] PROGMEM = "Reference_Lat";
static const char Config_Reference_Lon[] PROGMEM = "Reference_Lon";
static const char Config_Reference_LatF[] PROGMEM = "Reference_LatF";
static const char Config_Reference_LonF[] PROGMEM = "Reference_LonF";

static void Config_WriteString_P(
	const char *str,
	FIL        *file)
{
	char ch;

	while ((ch = pgm_read_byte(str++)))
	{
		f_putc(ch, file);
	}
}

static FRESULT Config_ReadSingle(
	const char *dir,
	const char *filename)
{
	size_t  len;
	char    *name;
	char    *result;
	int32_t val;
	float   valf;
	
	uint8_t first = 0;

	FRESULT res;

	res = f_chdir(dir);
	if (res != FR_OK) return res;
	
	res = f_open(&Main_file, filename, FA_READ);
	if (res != FR_OK) return res;

	bool ref_lat_valid = false;
	float ref_lat = 0.0;
	bool ref_lon_valid = false;
	float ref_lon = 0.0;

	while (!f_eof(&Main_file))
	{
		f_gets(UBX_buffer.buffer, sizeof(UBX_buffer.buffer), &Main_file);

		len = strcspn(UBX_buffer.buffer, ";");
		UBX_buffer.buffer[len] = 0;
		
		name = strtok(UBX_buffer.buffer, " \r\n\t:");
		if (name == 0) continue ;
		
		result = strtok(0, " \r\n\t:");
		if (result == 0) continue ;
		
		val = atol(result);
		valf = atof(result);
		
		#define HANDLE_VALUE(s,w,r,t) \
			if ((t) && !strcmp_P(name, (s))) { (w) = (r); }

		HANDLE_VALUE(Config_Model,     UBX_model,        val, val >= 0 && val <= 8);
		HANDLE_VALUE(Config_Rate,      UBX_rate,         val, val >= 100);
		HANDLE_VALUE(Config_Mode,      UBX_mode,         val, (val >= 0 && val <= 4) || (val == 11));
		HANDLE_VALUE(Config_Min,       UBX_min,          val, TRUE);
		HANDLE_VALUE(Config_Max,       UBX_max,          val, TRUE);
		HANDLE_VALUE(Config_Limits,    UBX_limits,       val, val >= 0 && val <= 2);
		HANDLE_VALUE(Config_Volume,    Tone_volume,      8 - val, val >= 0 && val <= 8);
		HANDLE_VALUE(Config_Mode_2,    UBX_mode_2,       val, (val >= 0 && val <= 4) || (val >= 8 && val <= 9) || (val == 11));
		HANDLE_VALUE(Config_Min_Val_2, UBX_min_2,        val, TRUE);
		HANDLE_VALUE(Config_Max_Val_2, UBX_max_2,        val, TRUE);
		HANDLE_VALUE(Config_Min_Rate,  UBX_min_rate,     val * TONE_RATE_ONE_HZ / 100, val >= 0);
		HANDLE_VALUE(Config_Max_Rate,  UBX_max_rate,     val * TONE_RATE_ONE_HZ / 100, val >= 0);
		HANDLE_VALUE(Config_Flatline,  UBX_flatline,     val, val == 0 || val == 1);
		HANDLE_VALUE(Config_Sp_Rate,   UBX_sp_rate,      val * 1000, val >= 0 && val <= 32);
		HANDLE_VALUE(Config_Sp_Volume, Tone_sp_volume,   8 - val, val >= 0 && val <= 8);
		HANDLE_VALUE(Config_V_Thresh,  UBX_threshold,    val, TRUE);
		HANDLE_VALUE(Config_H_Thresh,  UBX_hThreshold,   val, TRUE);
		HANDLE_VALUE(Config_Use_SAS,   UBX_use_sas,      val, val == 0 || val == 1);
		HANDLE_VALUE(Config_Window,    UBX_alarm_window_above, val * 1000, TRUE);
		HANDLE_VALUE(Config_Window,    UBX_alarm_window_below, val * 1000, TRUE);
		HANDLE_VALUE(Config_Window_Above, UBX_alarm_window_above, val * 1000, TRUE);
		HANDLE_VALUE(Config_Window_Below, UBX_alarm_window_below, val * 1000, TRUE);
		HANDLE_VALUE(Config_DZ_Elev,   UBX_dz_elev,      val * 1000, TRUE);
		HANDLE_VALUE(Config_TZ_Offset, Log_tz_offset,    val, TRUE);
		HANDLE_VALUE(Config_Init_Mode, UBX_init_mode,    val, val >= 0 && val <= 2);
		HANDLE_VALUE(Config_Alt_Units, UBX_alt_units,    val, val >= 0 && val <= 1);
		HANDLE_VALUE(Config_Alt_Step,  UBX_alt_step,     val, val >= 0);
		HANDLE_VALUE(Config_Acro_Win, UBX_acro_win, val * 1000, TRUE);
		HANDLE_VALUE(Config_Exit_Min_Alt,     UBX_exit_finder.min_alt_agl_mm,      val * 1000, TRUE);
		HANDLE_VALUE(Config_Exit_Exit_Alt,    UBX_exit_finder.cfg_exit_alt_agl_mm, val * 1000, TRUE);
		HANDLE_VALUE(Config_Exit_Num_Down,    UBX_exit_finder.num_down,            val,        val >= 0 && val <= 255);
		HANDLE_VALUE(Config_Exit_Num_Up,      UBX_exit_finder.num_up,              val,        val >= 0 && val <= 255);
		HANDLE_VALUE(Config_Exit_Down_Thresh, UBX_exit_finder.down_thresh_cmps,    val, TRUE);
		HANDLE_VALUE(Config_Exit_Up_Thresh,   UBX_exit_finder.up_thresh_cmps,      val, TRUE);

		HANDLE_VALUE(Config_Num_Blink_LEDs,   UBX_num_blink_leds,                  val, TRUE);
		HANDLE_VALUE(Config_Time_After_Exit,  UBX_nav.time_after_exit_ms,          val, TRUE);

		#undef HANDLE_VALUE
		
		if (!strcmp_P(name, Config_Init_File))
		{
			result[8] = '\0';
			strcpy(UBX_init_filename, result);
		}
		
		if (!strcmp_P(name, Config_Alarm_Elev) && UBX_num_alarms < UBX_MAX_ALARMS)
		{
			if (!(first & CONFIG_FIRST_ALARM))
			{
				UBX_num_alarms = 0;
				first |= CONFIG_FIRST_ALARM;
			}
			
			++UBX_num_alarms;
			UBX_alarms[UBX_num_alarms - 1].elev = val * 1000;
			UBX_alarms[UBX_num_alarms - 1].type = 0;
			UBX_alarms[UBX_num_alarms - 1].filename[0] = '\0';
		}
		if (!strcmp_P(name, Config_Alarm_Type) && UBX_num_alarms <= UBX_MAX_ALARMS)
		{
			UBX_alarms[UBX_num_alarms - 1].type = val;
		}
		if (!strcmp_P(name, Config_Alarm_File) && UBX_num_alarms <= UBX_MAX_ALARMS)
		{
			result[8] = '\0';
			strcpy(UBX_alarms[UBX_num_alarms - 1].filename, result);
		}
		if (!strcmp_P(name, Config_Alarm_Acro) && UBX_num_alarms <= UBX_MAX_ALARMS)
		{
			UBX_alarms[UBX_num_alarms - 1].acro_alarm = val;
		}
		
		if (!strcmp_P(name, Config_Wind_hMSL) && UBX_num_winds < UBX_MAX_WINDS)
		{
			if (!(first & CONFIG_FIRST_WIND))
			{
				UBX_num_winds = 0;
				first |= CONFIG_FIRST_WIND;
			}

			++UBX_num_winds;
			UBX_winds[UBX_num_winds - 1].hMSL = val;
			UBX_winds[UBX_num_winds - 1].velN = 0;
			UBX_winds[UBX_num_winds - 1].velE = 0;
		}
		if (!strcmp_P(name, Config_Wind_velN) && UBX_num_winds <= UBX_MAX_WINDS)
		{
			UBX_winds[UBX_num_winds - 1].velN = val;
		}
		if (!strcmp_P(name, Config_Wind_velE) && UBX_num_winds <= UBX_MAX_WINDS)
		{
			UBX_winds[UBX_num_winds - 1].velE = val;
		}

		if (!strcmp_P(name, Config_Win_Top) && UBX_num_windows < UBX_MAX_WINDOWS)
		{
			if (!(first & CONFIG_FIRST_WINDOW))
			{
				UBX_num_windows = 0;
				first |= CONFIG_FIRST_WINDOW;
			}
			
			++UBX_num_windows;
			UBX_windows[UBX_num_windows - 1].top = val * 1000;
		}
		if (!strcmp_P(name, Config_Win_Bottom) && UBX_num_windows <= UBX_MAX_WINDOWS)
		{
			UBX_windows[UBX_num_windows - 1].bottom = val * 1000;
		}

		if (!strcmp_P(name, Config_Sp_Mode) && UBX_num_speech < UBX_MAX_SPEECH)
		{
			if (!(first & CONFIG_FIRST_SPEECH))
			{
				UBX_num_speech = 0;
				first |= CONFIG_FIRST_SPEECH;
			}
			
			++UBX_num_speech;
			UBX_speech[UBX_num_speech - 1].mode = val;
			UBX_speech[UBX_num_speech - 1].units = UBX_UNITS_MPH;
			UBX_speech[UBX_num_speech - 1].decimals = 1;
		}
		if (!strcmp_P(name, Config_Sp_Units) && UBX_num_speech <= UBX_MAX_SPEECH)
		{
			UBX_speech[UBX_num_speech - 1].units = val;
		}
		if (!strcmp_P(name, Config_Sp_Dec) && UBX_num_speech <= UBX_MAX_SPEECH)
		{
			UBX_speech[UBX_num_speech - 1].decimals = val;
		}

		if (!strcmp_P(name, Config_Reference_Lat))
		{
			ref_lat_valid = true;
			ref_lat = val / LAT_LON_SCALE;
		}
		if (!strcmp_P(name, Config_Reference_LatF))
		{
			ref_lat_valid = true;
			ref_lat = valf;
		}

		if (!strcmp_P(name, Config_Reference_Lon))
		{
			ref_lon_valid = true;
			ref_lon = val / LAT_LON_SCALE;
		}
		if (!strcmp_P(name, Config_Reference_LonF))
		{
			ref_lon_valid = true;
			ref_lon = valf;
		}

		if (!strcmp_P(name, Config_Lane_Width))
		{
			UBX_pos_leds.set_lane_width(val);
		}
		if (!strcmp_P(name, Config_Lane_Num_LEDs))
		{
			UBX_pos_leds.set_num_leds(val);
		}
		if (!strcmp_P(name, Config_Lane_LED_Offset))
		{
			UBX_pos_leds.set_led_offset(val);
		}
	}

	if (ref_lat_valid && ref_lon_valid) {
		flysight::fw::LLA<float> ref_lla;
		ref_lla.latitude_deg = ref_lat;
		ref_lla.longitude_deg = ref_lon;
		ref_lla.height_msl_m = 0;
		UBX_nav.set_ref_lla(ref_lla);
	}
	
	f_close(&Main_file);
	
	return FR_OK;
}

void Config_Read(void)
{
	FRESULT res;

	res = Config_ReadSingle("\\", "config.txt");
	
	if (res != FR_OK)
	{
		res = f_chdir("\\");
		res = f_open(&Main_file, "config.txt", FA_WRITE | FA_CREATE_ALWAYS);
		if (res != FR_OK) 
		{
			Main_activeLED = LEDS_RED;
			UBX_ChangeLEDs(LEDS_ALL_LEDS, Main_activeLED);
			return ;
		}

		Config_WriteString_P(Config_default, &Main_file);
		f_close(&Main_file);
	}

	eeprom_read_block(UBX_buffer.filename, CONFIG_FNAME_ADDR, CONFIG_FNAME_LEN);

	if (UBX_buffer.filename[0] != 0 && UBX_buffer.filename[0] != 0xff)
	{
		res = Config_ReadSingle("\\config", UBX_buffer.filename);
	}
}
